### PR TITLE
Extend expression fuzzer test with PrestoQueryRunner.

### PR DESCRIPTION
**Pull request by kagamiori**
This PR adds the basic support of making expression fuzzer verify Velox results 
against Presto (https://github.com/facebookincubator/velox/issues/10308). 

This basic support has the following limitations that will be addressed in 
subsequent PRs:
1. Not testing arithmetic and comparison operators, e.g., +, -, <, >.
2. Not support dereference.
3. Not support literals of JSON, Timestamp, Interval, and complex types.
4. Not support AND/OR special forms.
5. The fuzzer currently ignores exceptions from reference DB when default-null 
behavior happens in Velox.

Differential Revision: D63009272

### CODE REVIEW

**Review by bikramSingh91 — COMMENTED** [review]
Just curious can you mention in https://github.com/facebookincubator/velox/issues/10308 what issues are causing the unsupported points from being enabled?

File: velox/exec/fuzzer/ReferenceQueryRunner.h
Comment (kagamiori) [reply velox/exec/fuzzer/ReferenceQueryRunner.h:61]: I think the criteria on valid function signatures depend on individual query runner, so it shouldn't be specified at this base class. I explained why function signatures using some types are not supported by PrestoQueryRunner in a comment in the PrestoQueryRunner::isSupported() implementation.

File: velox/exec/fuzzer/ToSQLUtil.cpp
Comment (kagamiori) [reply velox/exec/fuzzer/ToSQLUtil.cpp:194]: Agree. Json is supported though, through the branch of `else if (constant->type()->isPrimitiveType())`.

File: velox/expression/fuzzer/ExpressionFuzzer.cpp
Comment (kagamiori) [reply velox/expression/fuzzer/ExpressionFuzzer.cpp:834]: Good point. Will change the name. Thanks!

File: velox/expression/fuzzer/ExpressionFuzzer.cpp
Comment (kagamiori) [reply velox/expression/fuzzer/ExpressionFuzzer.cpp:920]: Good question. I don't think this cause a big change to the lambda expression coverage though. This change is basically saying making the lambda to be something like `(x, ...) -> c0` instead of the original `(x, ...) -> timestamp '1970-01-01 00:00:00'` when the ConstantTypedExpr is not supported. So for other supported ConstantTypedExpr, the lambda still returns the constant literal. Given that lambdas returning a constant literal are already trivial themselves, I don't think making the lambda returning a column instead would hurt the coverage.

The comment definitely should be updated.

File: velox/expression/fuzzer/ExpressionFuzzer.cpp
Comment (kagamiori) [reply velox/expression/fuzzer/ExpressionFuzzer.cpp:927]: No, "cast" has always been available. The problem is that originally we generate a CallTypedExpr with the name being "cast". But this doesn't work with PrestoQueryRunner when converting this CallTypedExpr to SQL because "cast" requires a different syntax from regular functions. So this PR makes it to generate a proper CastTypedExpr for "cast".

File: velox/expression/tests/ExpressionVerifier.cpp
Comment (kagamiori) [reply velox/expression/tests/ExpressionVerifier.cpp:210]: This is an old comment. I'll update it.

File: velox/exec/fuzzer/ToSQLUtil.cpp
Comment (kagamiori) [reply velox/exec/fuzzer/ToSQLUtil.cpp:160]: The logic are the same, but toConcatSql() cannot be merged into toCallSql() because ITypedExpr and ConcatTypedExpr don't have the name() method used in toCallSql(). So I think it's better to have them separated. Let me know if you have other suggestions.

File: velox/expression/fuzzer/ExpressionFuzzerVerifier.cpp
Comment (kagamiori) [reply velox/expression/fuzzer/ExpressionFuzzerVerifier.cpp:118]: Yeah, it somehow didn't work before I added this piece of code. But I don't see it in aggregation fuzzer. Let me dig deeper.

File: velox/expression/Expr.cpp
Comment (bikramSingh91) [inline velox/expression/Expr.cpp:1794]: isent the default value false? should this check be this instead?

`if (expr.stats().numProcessedRows || expr.stats().defaultNullRowsSkipped)`

Since we would want to know in case numProcessedRows is zero and defaultNullRowsSkipped=true helped skip all

File: velox/expression/Expr.h
Comment (bikramSingh91) [inline velox/expression/Expr.h:382]: is there any use case where we might want this? I wonder if we should just have an API in ConstantExpr class that would update the stat `defaultNullRowsSkipped`, and mention what its used for (only during constant folding).

Also, do we plan to remove all the logic around `defaultNullRowsSkipped` once presto and velox are in sync on this behavior?

File: velox/expression/fuzzer/ExpressionFuzzerTest.cpp
Comment (bikramSingh91) [inline velox/expression/fuzzer/ExpressionFuzzerTest.cpp:37]: can you also update the `fuzzer.rst' with details about these changes and the startup flags

File: velox/exec/fuzzer/FuzzerUtil.cpp
Comment (bikramSingh91) [inline velox/exec/fuzzer/FuzzerUtil.cpp:281]: do the children also need to be sanitized recursively?

File: velox/expression/Expr.h
Comment (bikramSingh91) [reply velox/expression/Expr.h:382]: since you have switched to using `setDefaultNullRowsSkipped`, should we get rid of this?

File: velox/expression/ExprCompiler.cpp
Comment (bikramSingh91) [inline velox/expression/ExprCompiler.cpp:317]: nice compaction

**Review by bikramSingh91 — APPROVED** [review]
Looks good. Accepting to unblock, but please address the two new comments that I added before merging. Thanks!

File: velox/exec/fuzzer/FuzzerUtil.cpp
Comment (kagamiori) [reply velox/exec/fuzzer/FuzzerUtil.cpp:281]: Nice catch. Thanks!

**Pull request closed without merging**
